src/server.rs

```diff
@@ -1,292 +1,305 @@
-use std::io::Read;
 use std::sync::Arc;
 use std::sync::RwLock;
 
-use byte_unit::Byte;
-use hyper::header::*;
-use jsonwebtoken::{decode, Validation};
-use nickel::{*, HttpRouter, Middleware, MiddlewareResult, Nickel, Options, Request, Response, StaticFilesHandler};
-use nickel::hyper::method::Method;
-use nickel::status::StatusCode;
+use bytes::Buf;
+use jsonwebtoken::Validation;
+use snafu::Snafu;
+use warp::{self, filters, fs, http::StatusCode, path, Filter, Rejection, Reply};
 
-use crate::configuration::*;
+use crate::configuration::{Claims, Configuration};
 use crate::kvstore::KvStore;
-use crate::logger::{Logger, LogLevel};
-
-pub struct Server {
-    configuration: Configuration
-}
 
 #[derive(Serialize, Deserialize)]
-struct ErrorMessage {
+struct JsonMessage {
     message: String,
 }
 
-#[derive(Serialize, Deserialize)]
-struct Claims {
-    sub: String,
-    iss: String,
-    iat: i64,
-    exp: i64,
+pub struct Server {
+    configuration: Arc<RwLock<Configuration>>,
 }
 
-#[derive(Serialize, Deserialize)]
-struct PatchValue {
-    operation: String
-}
+impl Server {
+    pub fn new() -> Server {
+        Server {
+            configuration: Arc::new(RwLock::new(Configuration::default())),
+        }
+    }
 
-fn middleware_webui<'a>(_: &mut Request, mut res: Response<'a>) -> MiddlewareResult<'a> {
-    res.set(MediaType::Html);
-    res.send_file("webui/dist/index.html")
-}
+    pub fn configure(&mut self, configuration: Configuration) {
+        *self.configuration.write().unwrap() = configuration;
+    }
 
-fn middleware_logging<'a, D>(request: &mut Request<D>, response: Response<'a, D>) -> MiddlewareResult<'a, D> {
-    crate::logger::print(&LogLevel::Information, format!("{} {}", request.origin.method, request.origin.uri).as_ref());
-    response.next_middleware()
-}
+    pub fn run(&self) {
+        let store = Arc::new(KvStore::new());
+        let store = warp::any().map(move || store.clone());
 
-fn middleware_cors<'mw>(_req: &mut Request, mut res: Response<'mw>) -> MiddlewareResult<'mw> {
-    res.headers_mut().set_raw("Access-Control-Allow-Origin", vec![b"*".to_vec()]);
-    res.headers_mut().set_raw("Access-Control-Allow-Methods", vec![b"*".to_vec()]);
-    res.headers_mut().set_raw("Access-Control-Allow-Headers", vec![b"*".to_vec()]); //Origin, Authorization, X-Requested-With, Content-Type, Accept
-    res.next_middleware()
-}
+        let config = self.configuration.clone();
+        let config = warp::any().map(move || config.clone());
 
-struct KvStoreMiddleware {
-    http_verb: hyper::method::Method,
-    store: Arc<RwLock<KvStore>>,
-    configuration: Configuration
-}
+        let auth = warp::header::optional::<String>("authorization")
```

Comment:
Could you enable JWT authentication only if it's enabled in the configuration?

Reply:
Yes, it should be easy to add a check if it's enabled in verify_auth and accept all requests if it's not.